app/org/sagebionetworks/bridge/hibernate/AccountPersistenceExceptionConverter.java

```
@@ -46,8 +47,18 @@ public RuntimeException convert(PersistenceException exception, Object entity) {
                 // "Duplicate entry 'api-ext' for key 'Accounts-StudyId-ExternalId-Index'" 
                 EntityAlreadyExistsException eae = null;
                 if (message.matches("Duplicate entry.*for key 'Accounts-StudyId-ExternalId-Index'")) {
-                    eae = createEntityAlreadyExistsException("External ID",
-                            AccountId.forExternalId(account.getStudyId(), account.getExternalId()));
+                    // We do not know which external ID is the conflict without parsing the error message. 
+                    // Try them until we find one. This external ID could be in a sub-study the caller is 
+                    // not associated to, but this is because external IDs have to be unique at the scope 
+                    // of the study, in order to identify substudy membership FROM the external ID.
+                    account = (HibernateAccount)BridgeUtils.filterForSubstudy(account);
+                    for (String externalId : BridgeUtils.collectExternalIds(account)) {
+                        eae = createEntityAlreadyExistsException("External ID",
+                                AccountId.forExternalId(account.getStudyId(), externalId));
+                        if (eae != null) {
+                            break;
+                        }
+                    }
```

What happens if none of the external IDs match? Do we give a different error message? Or do we fall back to the super-generic "Accounts table constraint prevented save or update"?

Do we have a unit test for this case?

---

We should throw a generic exception, because the caller is operating across subsidy boundaries and we don't want to leak information. Added test of this, and fixed bug in code when the passed in account does not in fact match the caller's substudies and is returned as null from BridgeUtils.filterForSubstudy().

---

Changed this to remove the filterForStubsudy(). It's too much security as you cannot hide the scope of external IDs between substudies (see comment above).